### PR TITLE
localCI: implements localCI

### PR DESCRIPTION
**Pull request by devimc**
localCI is a continuous integration system

Signed-off-by: Julio Montes <julio.montes@intel.com>

### CODE REVIEW

File: cmd/localCI/main.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+const name = "localCI"
+
+// usage is the usage of the program.
+const usage = name + ` is a continuous integration system for Intel® Clear Containers`
+
+var ciLog = logrus.New()
+
+func loop() {
+	var wg sync.WaitGroup
+
```
Comment (jodh-intel) [inline cmd/localCI/main.go]: I guess something is going to be slotted in here ;-)

**Comment by jodh-intel** [comment]
Hi @devimc - this looks interesting.

Note that you probably want to remove `cmd/localCI/localCI` which I assume was added by mistake?

File: cmd/localCI/config/configuration.toml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+[[github.com]]
+owner = "clearcontainers"
+repos = ["tests", "runtime", "proxy"]
```
Comment (dlespiau) [inline cmd/localCI/config/configuration.toml]: + "shim"!

Comment (dlespiau) [inline cmd/localCI/config/configuration.toml]: Can we have a nicer way to list the repos? If you want to use TOML, that's sounds fine to me, but let's make Repositories proper objects so we can add per-repository config options int the future.

You are using an "array of table" which is really an array of objects. You're objects are a bit off here, what about:

```
[[Repositories]]
name = runtime
url = https://github.com/...
test_action = make check
...

[[Repositories]]
name = proxy
url = https://github.com/...

...

[[Repositories]]
name = ...
url = ...
```

**Comment by devimc** [comment]
@jodh-intel yes, I'll remove ```localCI/localCI```

File: cmd/localCI/main.go (hunk 1 above)
Comment (devimc) [inline cmd/localCI/main.go]: yes, I plan to launch go routines to monitor each repository

File: cmd/localCI/config/configuration.toml (hunk 2 above)
Comment (devimc) [inline cmd/localCI/config/configuration.toml]: yes, that looks better

File: cmd/localCI/Makefile [hunk 3]
```diff
@@ -0,0 +1,24 @@
+DESTDIR :=
+PREFIX := /usr/local
+BINDIR := $(PREFIX)/bin
+SYSCONFDIR := $(PREFIX)/etc
+LIBEXECDIR := $(PREFIX)/libexec
+
+SOURCES := $(shell find . 2>&1 | grep -E '.*\.(c|h|go)$$')
```
Comment (grahamwhaley) [inline cmd/localCI/Makefile]: Just a note - I'm not a fan of such wildcarding in Makefiles - at some point somebody (usually me ;-) will end up copying a file whilst experimenting, and then spend ages trying to figure out why the system now won't build or fails or ..... due to the wildcarding picking up files that are not actually part of the project.

Comment (dlespiau) [inline cmd/localCI/Makefile]: It's also not needed, just calling `go build` unconditionally is perfectly fine.

Comment (jodh-intel) [inline cmd/localCI/Makefile]: It's worth noting that `go build` will essentially do the same as the `find` though: if you create a new `.go` file as an experiment, `go build` **will** try to build it ;-)

Comment (dlespiau) [inline cmd/localCI/Makefile]: which is perfectly fine :)

Note re-triggering the build depends on what you do. I usually add `-i` to `go build` so the the dependent packages are installed. That makes `go build` skip the compilation of those packages if no file is newer than the installed package.

For instance, compiling the ./api package in the proxy (which uses `go build -i`):

```
$ go build -x ./api/
WORK=/tmp/go-build703241415
```

Nothing!

Given the speed of go build and the tracking of installed packages, we don't gain anything (or much) depending on the files in the Makefile rune. I'd would just get rid of it.

**Comment by grahamwhaley** [comment]
@devimc It will take me a bit of time to review the code drop (~1700+ lines in 14 files or so, without the vendor tree), but if you 'bored and wanting something to do', then I did run the `.ci/go-lint.sh` on it, and it threw out a few things. I will note though, the `go-static-checks.sh` was nice and clean ;-)

**Comment by grahamwhaley** [comment]
Oh, if you do run `go-lint`, watch out for it reporting all the vendor stuff - I don't know why, it looks like it is configured not to - but, it still seems it does. Maybe somebody who understands vendoring can see if that is our issue or a problem with the gometalinter?

**Comment by devimc** [comment]
@grahamwhaley yep go-lint reports some warnings

File: cmd/localCI/README.rst [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+- ``useVM [bool] (optional)``: specify if **localCI** should use virtual machines to test the pull requests. By default this value is ``false``.
+- ``logDir [string] (optional)``: the directory where the logs are placed. By default this value is ``/var/log/localCI``
+
+and have next sub-sections
+
+- ``[Repo.language] (mandatory)``
+
+  The language of the project.
+
+  - ``language [string] (mandatory)``: the language of the project. To check the list of supported languages see `Supported languages`_
```
Comment (grahamwhaley) [inline cmd/localCI/README.rst]: Just a passing comment - in the future maybe we will need to support multiple languages per repo for mixed language repos.

File: cmd/localCI/README.rst [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+  - ``version [string] (madatory if 'language = Go')``: the version of the ``language`` that must be used to test the project.
+
+- ``[Repo.commentTrigger] (optional)``
+
+  Comment nedded in the pull request to test it.
```
Comment (grahamwhaley) [inline cmd/localCI/README.rst]: s/nedded/needed/

File: cmd/localCI/README.rst [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+..........................
+::
+
+   runTestsInParallel = true
+
+   [[Repo]]
+   url = "https://github.com/clearcontainers/runtime"
+   pr = 563
+   refreshTime = "60s"
+   token = "Abccvf457sdm89mous756sx"
```
Comment (grahamwhaley) [inline cmd/localCI/README.rst]: I'm presuming this is not a real live token ;-)

**Comment by grahamwhaley** [comment]
With feedback left:
lgtm

File: cmd/localCI/README.rst [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+     language = "Go"
+     version = "go1.8.3"
+   [Repo.commentTrigger]
+     user = "QA-bot"
+     comment = "qa-passed"
+   [Repo.logServer]
+	 ip = "192.168.1.15"
+	 user = "root"
+	 dir = "/var/log/localCI"
+	 publicKey = "ssh-rsa AAAAAAAnmcjdNoNksmUus84"
```
Comment (devimc) [inline cmd/localCI/README.rst]: yes, it is just an example

File: cmd/localCI/README.rst (hunk 6 above)
Comment (devimc) [inline cmd/localCI/README.rst]: it is a fake token

File: cmd/localCI/README.rst (hunk 5 above)
Comment (devimc) [inline cmd/localCI/README.rst]: fixing, thanks

File: cmd/localCI/README.rst [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+The commans of each stage run sequentially, inherit the environment variables of the system and some others
+`Environment variables`_ are exported.
+
+Logging
+-------
+
+By default **localCI** places all the logs generated by each stage in ``/var/log/localCI/`` inside a directory with the
+same number of the pull request, but you can specify a different path modifying ``logDir`` in the `Configuration file`_.
+Two files with the suffix ``.stdout`` and ``.stderr`` are generated by stage. For example if the ``logDir`` is
+``/var/log/localCI`` and the pull pull request that is being tested is the number ``53``, when the stage ``setup`` runs
```
Comment (devimc) [inline cmd/localCI/README.rst]: yep, fixing

File: cmd/localCI/README.rst [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
++---------------------+----------------------+-----------------------------------------------------+
+| LOCALCI             | true                 | The CI is **localCI**                               |
++---------------------+----------------------+-----------------------------------------------------+
+| LOCALCI_REPO_SLUG   | E.g foo/bar          | Repository owner / repository name                  |
++---------------------+----------------------+-----------------------------------------------------+
+
+
+Configuration file
+------------------
+The configuration file follows the syntax defined by the `Tom's Obvious, Minimal Language`_ and might contain next keys and sections:
```
Comment (devimc) [inline cmd/localCI/README.rst]: good catch
thanks

File: cmd/localCI/README.rst [hunk 10]
```diff
@@ -0,0 +1,156 @@
+localCI
+=======
+
+LocalCI is a Continuous Integration System designed to have two way of use
```
Comment (devimc) [inline cmd/localCI/README.rst]: the second one sounds better
thanks

File: cmd/localCI/repo.go [hunk 11]
```diff
@@ -0,0 +1,398 @@
...
+			ciLog.Errorf("failed to run stage '%s': %s", s.name, err)
+			if len(r.PostOnFailure) == 0 {
+				continue
+			}
+
+			if err = r.cvr.createComment(pr.Number, r.PostOnFailure); err != nil {
+				return fmt.Errorf("failed to create comment '%s' on pull request %d", r.PostOnFailure, pr.Number)
+			}
+
+			_ = pr.runStage("onFailure", r.onFailure)
```
Comment (devimc) [inline cmd/localCI/repo.go]: yes, I can add an error message

File: cmd/localCI/pullRequest.go [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+	for i, c := range pr.Commits {
+		if strings.Compare(c.Sha, rpr.Commits[i].Sha) != 0 {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (pr *PullRequest) runStage(stage string, commands []string) error {
```
Comment (devimc) [inline cmd/localCI/pullRequest.go]: good catch !

File: cmd/localCI/repo.go [hunk 13]
```diff
@@ -0,0 +1,398 @@
...
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+// Repo represents the repository under test
+type Repo struct {
+	URL            string
+	PR             int
```
Comment (devimc) [inline cmd/localCI/repo.go]: fixing

File: cmd/localCI/pullRequest.go [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+// canBeTested returns an error if the pull request cannot be tested
+func (pr *PullRequest) canBeTested() error {
+	commitsLen := len(pr.Commits)
+	if commitsLen == 0 {
+		return fmt.Errorf("there are no commits to test")
+	}
+
+	// there is no a comment trigger so we can test
```
Comment (devimc) [inline cmd/localCI/pullRequest.go]: fixing, thanks

File: cmd/localCI/main.go [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+		}
+
+		return nil
+	}
+
+	app.Action = func(context *cli.Context) error {
+		repos := context.App.Metadata["repos"].([]Repo)
+
+		if len(repos) == 0 {
+			ciLog.Info("There is no repos to monitor or test")
```
Comment (devimc) [inline cmd/localCI/main.go]: good catch! fixing

File: cmd/localCI/go_test.go [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+	for _, e := range goLanguage.getEnv() {
+		if strings.HasPrefix(e, "GOROOT") {
+			noGoRoot = false
+		} else if strings.HasPrefix(e, "GOPATH") {
+			noGopath = false
+		}
+	}
+	assert.False(noGopath)
+	assert.False(noGoRoot)
+}
```
Comment (devimc) [inline cmd/localCI/go_test.go]: nop, because I'm no testing (not yet) the setup method

File: cmd/localCI/github.go [hunk 17]
```diff
@@ -0,0 +1,240 @@
...
+	defer cancel()
+
+	pullRequest, _, err := g.client.PullRequests.Get(ctx, g.owner, g.repo, pr)
+	if err != nil {
+		return nil, err
+	}
+
+	isOpen := false
+	if strings.Compare(*pullRequest.State, "open") == 0 {
+		isOpen = true
```
Comment (devimc) [inline cmd/localCI/github.go]: yes, I think we do not need this field

File: cmd/localCI/github.go [hunk 18]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+// getRepo returns the repository name
+func (g *Github) getRepo() string {
+	return g.repo
+}
+
+// getOpenPullRequests returns the open pull requests
+func (g *Github) getOpenPullRequests() (map[string]*PullRequest, error) {
+	// maximum 5 seconds to complete the request
```
Comment (devimc) [inline cmd/localCI/github.go]: sure, fixing

File: cmd/localCI/README.rst [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+
+The commans of each stage run sequentially, inherit the environment variables of the system and some others
+`Environment variables`_ are exported.
+
+Logging
+-------
+
+By default **localCI** places all the logs generated by each stage in ``/var/log/localCI/`` inside a directory with the
+same number of the pull request, but you can specify a different path modifying ``logDir`` in the `Configuration file`_.
+Two files with the suffix ``.stdout`` and ``.stderr`` are generated by stage. For example if the ``logDir`` is
```
Comment (devimc) [inline cmd/localCI/README.rst]: yep, setup, run, teardown ....

File: cmd/localCI/README.rst (hunk 4 above)
Comment (devimc) [inline cmd/localCI/README.rst]: sure , I'll keep it in mind

**Comment by grahamwhaley** [comment]
whitelist additions:
lgtm
I'll note I didn't see any unit tests, but they might be a little tricky as you'd have to rely on data from github itself or would have to make a github mock I guess.

**Comment by grahamwhaley** [comment]
Oh, I'll also note that the whitelisting feature probably fixes/replaces #129

File: cmd/localCI/vendor/github.com/golang/protobuf/README.md [hunk 20]
```diff
@@ -0,0 +1,199 @@
...
+Once the software is installed, there are two steps to using it.
+First you must compile the protocol buffer definitions and then import
+them, with the support library, into your program.
+
+To compile the protocol buffer definition, run protoc with the --go_out
+parameter set to the directory you want to output the Go code to.
+
+	protoc --go_out=. *.proto
+
+The generated files will be suffixed .pb.go.  See the Test code below
```
Comment (iphutch) [inline cmd/localCI/vendor/github.com/golang/protobuf/README.md]: Add code block (```) for the above command. When missing, everything that follows the * is italicized.

File: cmd/localCI/vendor/github.com/google/go-github/CONTRIBUTING.md [hunk 21]
```diff
@@ -0,0 +1,114 @@
+# How to contribute #
+
+We'd love to accept your patches and contributions to this project. There are
+a just a few small guidelines you need to follow.
```
Comment (iphutch) [inline cmd/localCI/vendor/github.com/google/go-github/CONTRIBUTING.md]: remove "a"

File: cmd/localCI/vendor/github.com/google/go-querystring/README.md [hunk 22]
```diff
@@ -0,0 +1,39 @@
+# go-querystring #
+
+go-querystring is Go library for encoding structs into URL query parameters.
```
Comment (iphutch) [inline cmd/localCI/vendor/github.com/google/go-querystring/README.md]: go-querystring is a Go library. "a" missing.

File: cmd/localCI/vendor/golang.org/x/oauth2/README.md [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+
+oauth2 package contains a client implementation for OAuth 2.0 spec.
+
+## Installation
+
+~~~~
+go get golang.org/x/oauth2
+~~~~
+
+See godoc for further documentation and examples.
```
Comment (iphutch) [inline cmd/localCI/vendor/golang.org/x/oauth2/README.md]: replace "godoc" with "godoc.org"

File: cmd/localCI/README.rst [hunk 24]
```diff
@@ -0,0 +1,160 @@
...
+
+**localCI** defines 5 stages when a pull request is being tested
+
+- ``Setup`` - The commands needed to setup the environment
+- ``Run`` - The commands needed to run the test
+- ``Teardown`` - The commands needed to cleanup the environment
+- ``On Success`` - The commands to be executed if ``Setup`` , ``Run`` and ``Teardown`` finished correctly
+- ``On Failure`` - The commands to be executed if any of ``Setup`` , ``Run`` or ``Teardown`` fail.
+
+The commans of each stage run sequentially, inherit the environment variables of the system and some others
```
Comment (devimc) [inline cmd/localCI/README.rst]: good catch!  thanks

File: cmd/localCI/vendor/github.com/BurntSushi/toml/cmd/toml-test-decoder/README.md [hunk 25]
```diff
@@ -0,0 +1,13 @@
+# Implements the TOML test suite interface
+
+This is an implementation of the interface expected by
+[toml-test](https://github.com/BurntSushi/toml-test) for my
```
Comment (devimc) [inline cmd/localCI/vendor/github.com/BurntSushi/toml/cmd/toml-test-decoder/README.md]: Hi @iphutch  thanks for reviewing, the ```vendor``` directory contains third party code that we can't modify

**Comment by devimc** [comment]
@iphutch @grahamwhaley @chavafg  can we merge this PR?

**Comment by grahamwhaley** [comment]
@devimc I approved from the code pov. You still have CI broken it looks like (Travis).
It would be nice (if you have not already - I'm not in a position to check right now) if you managed to clean up those golint warnings.
Once CI is happy and we have docs signoff then I believe we can merge. I expect us to then have some on-going additions/improvements as we go through integration.

**Comment by devimc** [comment]
@grahamwhaley thanks

**Pull request merged**